packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -806,36 +806,39 @@ export abstract class FluidDataStoreContext
 			}
 
 			this.thresholdOpsCounter.send("ProcessPendingOps", pending.length);
-			this.pending = undefined;
+		} else {
+			assert(this.pending?.length === 0, "no pending ops");
+
+			// Load the handle to the data store's entryPoint to make sure that for a detached data store, the entryPoint
+			// initialization function is called before the data store gets attached and potentially connected to the
+			// delta stream, so it gets a chance to do things while the data store is still "purely local".
+			// This preserves the behavior from before we introduced entryPoints, where the instantiateDataStore method
+			// of data store factories tends to construct the data object (at least kick off an async method that returns
+			// it); that code moved to the entryPoint initialization function, so we want to ensure it still executes
+			// before the data store is attached.
+			await channel.entryPoint.get();
+		}
+		this.pending = undefined;
 
-			// And now mark the runtime active
-			this.loaded = true;
-			this.channel = channel;
+		// And now mark the runtime active
+		this.loaded = true;
+		this.channel = channel;
 
-			// Freeze the package path to ensure that someone doesn't modify it when it is
-			// returned in packagePath().
-			Object.freeze(this.pkg);
+		// Freeze the package path to ensure that someone doesn't modify it when it is
+		// returned in packagePath().
+		Object.freeze(this.pkg);
 
-			/**
-			 * Update the used routes of the channel. If GC has run before this data store was realized, we will have
-			 * the used routes saved. So, this will ensure that all the child contexts have up-to-date used routes as
-			 * per the last time GC was run.
-			 * Also, this data store may have been realized during summarize. In that case, the child contexts need to
-			 * have their used routes updated to determine if its needs to summarize again and to add it to the summary.
-			 */
-			this.updateChannelUsedRoutes();
+		/**
+		 * Update the used routes of the channel. If GC has run before this data store was realized, we will have
+		 * the used routes saved. So, this will ensure that all the child contexts have up-to-date used routes as
+		 * per the last time GC was run.
+		 * Also, this data store may have been realized during summarize. In that case, the child contexts need to
+		 * have their used routes updated to determine if its needs to summarize again and to add it to the summary.
+		 */
+		this.updateChannelUsedRoutes();
 
-			// And notify the pending promise it is now available
-			this.channelDeferred.resolve(this.channel);
-		} catch (error) {
-			this.channelDeferred?.reject(error);
-			this.mc.logger.sendErrorEvent(
-				{
-					eventName: "BindRuntimeError",
-				},
-				error,
-			);
-		}
+		// And notify the pending promise it is now available
+		this.channelDeferred.resolve(this.channel);
```

Comment:
Don't we want to reject the channelDeferred promise if something in this function fails?

Reply:
We do. Both callers of bindRuntime() (see attachRuntime & realize()) will do so.

Before my change try/catch was here, but realize() was duplicating try/catch, was wrapping error with DataProcessingError.wrapIfUnrecognized, but that wrapping was useless, as this.channelDeferred was storing unwrapped error and was used to return an error to user. Plus attachRuntime() was eating this exception, which is really bad.

With my change, exception is thrown from this function - to try/catch.

I agree that setting this.channelDeferred here looks somewhere weird - I'll remove it and move it to those to callers - that makes more sense - code would be more symmetrical RE where this.channelDeferred is handled. I also do not want to have try/catch here, as try/catch in callers captures bigger block of code.